npm/vite-plugin-cypress-esm/client/moduleCache.js

```diff
@@ -0,0 +1,145 @@
+const __cypressModuleCache = new Map()
+
+const NO_REDEFINE_LIST = new Set(['prototype'])
+
+function createProxyModule (module) {
+  // What we build our module proxy off of depends on whether the module has a default export
+  // We need to be able to support `import DefaultValue from 'module'` => `const DefaultValue = __cypressModule(module)`
+  const base = module.default || module
+  let target
+
+  // Work around for the fact that a module with a default export needs to work the same way via object destructuring
+  // for this module remapping concept to work
+  // ```
+  // import TheDefault from 'module'
+  // `TheDefault` could be an object or a function
+  // ```
+  if (typeof base === 'function') {
+    target = function (...params) {
+      if (typeof target.default === 'function') {
+        return target.default.apply(this, params)
+      }
+
+      if (typeof module === 'function') {
+        return module.apply(this, params)
+      }
+    }
+  } else {
+    target = {}
+  }
+
+  const proxies = {}
+
+  const redefinePropertyDescriptors = (module, overrides) => {
+    Object.entries(Object.getOwnPropertyDescriptors(module)).forEach(([key, descriptor]) => {
+      if (NO_REDEFINE_LIST.has(key)) {
+        return
+      }
+
+      Object.defineProperty(target, key, {
+        ...descriptor,
+        ...overrides,
+      })
+
+      if (typeof descriptor.value === 'function') {
+        proxies[key] = function (...params) {
+          return target[key].apply(this, params)
+        }
+      }
+    })
+  }
+
+  if (module.default && typeof module.default !== 'function') {
+    redefinePropertyDescriptors(module.default, {
+      writable: true,
+      enumerable: true,
+    })
+  }
+
+  redefinePropertyDescriptors(module, {
+    configurable: true,
+    writable: true,
+  })
+
+  const moduleProxy = new Proxy(target, {
+    get (_, prop, receiver) {
+      const value = target[prop]
+
+      if (typeof value === 'function') {
+        const stack = new Error().stack
+
+        if (stack.includes('Sandbox.spy')) {
+          return value
+        }
+
+        return proxies[prop]
+      }
+
+      return target[prop]
+    },
+    set (obj, prop, value) {
+      target[prop] = value
+
+      if (typeof value === 'function' && !(prop in proxies)) {
+        proxies[prop] = function (...params) {
+          return target[prop].apply(this, params)
+        }
```

Comment:
nitpick: This duplicates the function wrapping we do on line 45, could pull out a `wrapFunction` helper just to make sure they stay aligned

Reply:
Now that I think about it I can't remember why I found this block necessary. Theoretically all functions *should* have been proxied up front 🤔

Reply:
I tried to make this refactor, it's quite awkward and I think it's a little less clear to read. Any recommendation for how you think it could be better written?

Reply:
Just wondering now that we have more custom logic in this area if we'll need to apply it (eg constructor handling) in other places we're building function wrappers - there's like 4 places we're generating wrappers, hard to keep track of which ones need to account for what.